packages/NodeCollector/BinaryOpTreeRootLocator.php

```diff
@@ -20,17 +23,22 @@ final class BinaryOpTreeRootLocator
      */
     public function findOperationRoot(Expr $expr, string $binaryOpClass): Expr
     {
-        /** @var ?Expr $parentNode */
         $parentNode = $expr->getAttribute(AttributeKey::PARENT_NODE);
-        if ($parentNode === null || $parentNode::class !== $binaryOpClass) {
+        if (! $parentNode instanceof Expr) {
+            return $expr;
+        }
+        if ($parentNode::class !== $binaryOpClass) {
             return $expr;
         }
 
         assert($parentNode instanceof BinaryOp);
         $isLeftChild = $parentNode->left === $expr;
         $isRightChild = $parentNode->right === $expr;
         $isRightChildAndNotBinaryOp = $isRightChild && $expr::class !== $binaryOpClass;
-        if ($isLeftChild || $isRightChildAndNotBinaryOp) {
+        if ($isLeftChild) {
+            return $this->findOperationRoot($parentNode, $binaryOpClass);
+        }
+        if ($isRightChildAndNotBinaryOp) {
```

Comment:
I also do not understand its obsession with duplicating conditions like this.

if it really wants a single value in condition it could instead change it to:

```php
$condition = $isLeftChild || $isRightChildAndNotBinaryOp;
if ($condition) {
    return $this->findOperationRoot($parentNode, $binaryOpClass);
}
```

Reply:
This place can be improved :+1:

Reply:
Ideally in early return with cheap return first.

> if it really wants a single value in condition it could instead change it to:

That's one of ways :+1: 

The reason is to avoid piling up pipes ` || ` and require single condition with early returns.